### PR TITLE
Stop using iter_all_segments in Search, remove iter_all_segments API

### PR DESCRIPTION
**Pull request by SkySkimmer**
overlay: 
- https://github.com/coq-community/coq-dpdgraph/pull/107

### CODE REVIEW

**Comment by ejgallego** [comment]
Note that in principle I like this, in the sense that I like we use a "single source of truth" about what objects we know.

But as I noted in #16261 , this approach would have Search to query a new `LocTab` table, which may not be in sync with the kernel env automatically now.

**Comment by SkySkimmer** [comment]
Locate has nothing to do with this
The objects search looks at are not especially more in sync with the kernel than a table would be

**Comment by ejgallego** [comment]
> Locate has nothing to do with this

Fair point, tho still uses some libobject internals, but that's too wide in scope here.

> The objects search looks at are not especially more in sync with the kernel than a table would be

Indeed that's a bit orthogonal.

As I said I like this, but still I'm unsure if we want to restrict Search to only things the kernel know, or if we are going then to introduce N databases for N different objects.

I am positive I want to search in my interfaces for stuff that is outside of the kernel, for example Hint declarations.

**Comment by SkySkimmer** [comment]
>I think this needs discussion.

This PR is just removing a hack, what's there to discuss?

**Comment by ejgallego** [comment]
This:

> As I said I like this, but still I'm unsure if we want to restrict Search to only things the kernel know, or if we are going then to introduce N databases for N different objects.
> 
> I am positive I want to search in my interfaces for stuff that is outside of the kernel, for example Hint declarations.

In particular, I'm still not sure that iterating the libobject is the wrong approach here, on in other words, why is the current approach a hack?

**Comment by SkySkimmer** [comment]
>    As I said I like this, but still I'm unsure if we want to restrict Search to only things the kernel know, or if we are going then to introduce N databases for N different objects.

Libobject is an implementation detail to persist stuff across modules and sections, not a database. 

Iterating over libobject is nonsense. If search had originally been implemented the way it is in this PR we would never merge a PR making it the way it is in master.

**Comment by SkySkimmer** [comment]
See also the way we had to have a dummy object for variables because of using this method for search.

**Comment by ejgallego** [comment]
> Libobject is an implementation detail to persist stuff across modules and sections, not a database.

I'm not sure I see how these two definitions are in conflict, but I still think discussing this in a call would be more efficient.

I agree libobject API is very bad these days (and declaremods needs refactor etc)

> Iterating over libobject is nonsense. If search had originally been implemented the way it is in this PR we would never merge a PR making it the way it is in master.

Maybe we would not have, but my key question is:
- for search we need to access something alike to a database (kernel envs and module contents are both not a database but could fill the role)
- if we choose kernel envs as our database, how do we search for "non-logical" objects in scope, such as notations or hints?

**Comment by SkySkimmer** [comment]
>if we choose kernel envs as our database, how do we search for "non-logical" objects in scope, such as notations or hints?

Not with libobject, which is all that's relevant for this PR.

**Comment by ejgallego** [comment]
> Not with libobject, which is all that's relevant for this PR.

Why not with libobject? What would we use then?

After thinking a bit about it, I think it makes a lot of sense to expose module contents to clients (such as Search or Locate) with a nice API. I am not sure I get why that is "bad".

The alternative of having n-tables of objects seems pretty problematic to me, and with current infra pretty hard to maintain / brittle.

It is true that we should be careful in terms of not exposing implementation details, but having N different databases is IMHO more leaky than a single, well-designed one. But on the contrary, module contents have pretty good properties client side, first one is that they are extensible so that makes client maintenance way more robust IMO.

I am writing a kind of CEP tho to outline what I have in mind after my last experiments.

**Comment by ejgallego** [comment]
> See also the way we had to have a dummy object for variables because of using this method for search.

That's a good question, the way I understand it is "should every 'object' present in a module have a corresponding entry in the module contents table" ?

**Comment by SkySkimmer** [comment]
>That's a good question, the way I understand it is "should every 'object' present in a module have a corresponding entry in the module contents table" ?

What is an object? Do local things count? I'm not sure if it's still there but at some point we had code like `if local then cache obj else add_leaf obj`

I really think libobject is strictly about persistence, if we want to know what currently exists we must look at the summary.

**Comment by ejgallego** [comment]
> What is an object? Do local things count? I'm not sure if it's still there but at some point we had code like `if local then cache obj else add_leaf obj`

Yes, current semantics are quite fuzzy and need improvement. So maybe we just need a different thing and retro-fitting module contents is not the way to go. For now my typical use cases are "what hints do a module / library define", etc...

Actually in a sense the way I would envision something like this would be actually splitting concerns in the current implementation fo the module system. Persistence is important, but should be more orthogonal IMHO.

> I really think libobject is strictly about persistence, if we want to know what currently exists we must look at the summary.

Problem with Summary is that it also includes quite a bit of internal state that is not relevant to things in .v files, so I dunno.

In fact, it seems to me that it is usually the module implementation (aka libobject, but really declaremods) which is the one driving summary declarations. Moreover, we really have no guarantees about objects adding stuff to summary.

It seems to me that Summary is more about "state" not about "module contents" which what I care about.

For example, I am positive that we could remove quite a bit of boilerplate would we add some API such as `Libobject.declare_boolean_object` or `Libobject.declare_map_object`, where these methods are instrumented to declare the corresponding summary entries. Same for "objects" with location (then the implementation of the location is hidden from the client), or with names.

So more generically, I'd like we have:
- a (hopefully generic) way to query for the things that a module defines (that's pretty linked to persistence, but indeed not the same)
- a (hopefully generic) way to store metadata for such objects, such as documentation, location, whether it was defined by the user or not, name if the object has one, etc...

Another small thing is that I've used `iter_all_segments` in the past to actually dump what Coq knows to be in the .vo files, and indeed I think that's a need devs often may have.

File: vernac/search.ml
Comment (maximedenes) [inline vernac/search.ml:93]: Do you want to resolve this TODO as part of this PR, since the calls in the `try` block have changed?

**Review by maximedenes — COMMENTED** [review]
Nice clean-up!

**Comment by maximedenes** [comment]
I think it is fairly obvious that reusing libobject for search is a hack.

As for the need of dumping "what Coq knows to be in the .vo file", can you make the use case more specific? Why is e.g. votour not sufficient? Anyway such debugging tools should go in dedicated places (like debug printers) IMHO, not the main API.

**Comment by ejgallego** [comment]
I think this approach is not the best one, I'm working to summarize all the reasons and produce an alternative design as a CEP, a bit of discussion is already on #16261 .

We had quite a discussion on the CoqDB workshop, there are notes and a video about that too (I think you couldn't attend?)

@SkySkimmer I'm not sure why you did remove the `needs: discussion` tag unilaterally when we didn't discuss about it. But indeed, I am formally requesting not to merge this PR until we discuss alternative approaches.

Next Coq call would be good for me (the 7th I'm teaching so we'd have to the 14th)

**Comment by ejgallego** [comment]
Very briefly:
- I agree the current approoach is hack
- I don't agree the right solution is to rely on the kernel, as it doesn't scale properly for stuff that's not handled by the kernel.

So indeed I'd like to propose a refactoring of Coq's module system itself that accounts for all kind of objects in modules, separates some concerns, and provides a good client experience.

**Comment by maximedenes** [comment]
Why not first remove the hack (which is what this PR does)? Then of course you can implement a mechanism that scales better (i.e. is more general) than using the kernel (which is fine for the search feature as it is today).

I don't see the point of putting on hold an iso feature clean-up on which the main maintainers of the component agree, for the sake of discussing alternative mechanisms which can be implemented afterwards anyway.

**Comment by SkySkimmer** [comment]
>I don't agree the right solution is to rely on the kernel, as it doesn't scale properly for stuff that's not handled by the kernel.

I don't agree that scaling to hypothetical features is something we should care about, at least here where we're not changing the search API, nothing depends on how its internals work and so it should be easy to change once something actually needs it.

**Comment by ejgallego** [comment]
Note that I didn't block the PR, I was just requesting a discussion, so still it is very puzzling to me that tag was removed unilaterally.

**Comment by maximedenes** [comment]
The problem with discussing how we can do more *before* cleaning-up code is that it does not scale. If we do that for every feature we may want on every component, I'm afraid we will never reach the cleaning part. And the code base will be less maintainable, changes will be more costly, etc.

Speaking for myself, I unfortunately won't have time to discuss new search features before finishing current work on VsCoq. But I would certainly benefit (like other developers from the team) from cleaner libobject foundations (hence some previous clean-ups I already did there). So this PR is a net gain for me.

**Comment by maximedenes** [comment]
> Note that I didn't block the PR, I was just requesting a discussion, so still it is very puzzling to me that tag was removed unilaterally.

Doesn't this label mean "block the PR until some discussion takes place"?

**Comment by Zimmi48** [comment]
>Doesn't this label mean "block the PR until some discussion takes place"?

Absolutely since `needs:` label do actually prevent from merging a PR.

**Comment by ejgallego** [comment]
> Doesn't this label mean "block the PR until some discussion takes place"?

In that sense yes, what I mean is that it is not a "request for changes".

> The problem with discussing how we can do more _before_ cleaning-up code is that it does not scale. If we do that for every feature we may want on every component, I'm afraid we will never reach the cleaning part. And the code base will be less maintainable, changes will be more costly, etc.

Yes, it is clear discussing every little change can't scale. However here I want to discuss a specific cleanup that breaks my PR, because indeed I think we are going on the wrong technical path. So indeed, what do you suggest a maintainer to do when such maintainer thinks a change is not going on the right direction, if not to discuss it?

> Speaking for myself, I unfortunately won't have time to discuss new search features before finishing current work on VsCoq. But I would certainly benefit (like other developers from the team) from cleaner libobject foundations (hence some previous clean-ups I already did there). So this PR is a net gain for me.

How is the PR a net gain for you? It doesn't touch any libobject API, does it?

But yes, the PR may be a net gain for you, but is a problem for me, so hence I need to discuss how to move forward with my own PR (prior to this one) if this one is merged.

**Comment by maximedenes** [comment]
> How is the PR a net gain for you? It doesn't touch any libobject API, does it?

I'm actually working at the declaremods level, where this PR removes an API which use to leak libobject implementation details.

**Comment by maximedenes** [comment]
Anyway, I don't care much since I also can do with it, the point was just to illustrate.

Having expressed my opinion and exhausted the time I could spend on it, I'll let someone else handle it.

**Review by maximedenes — APPROVED** [review]
Looks good to me.

**Comment by gares** [comment]
OK, since the PR is not blocked I'll merge it once CI is over.

**Comment by ejgallego** [comment]
> OK, since the PR is not blocked I'll merge it once CI is over.

Have you read the discussion? Is there some part that is not clear, if so please let me know.

I think I have made it very clear I'd like to discuss before we merge this. And I did setup the tag which was removed unilaterally.

As I wrote, I didn't mean "not blocked", but "I don't necessarily have a request for changes".

**Comment by SkySkimmer** [comment]
>So again I don't see the harm in discussing what to do; and indeed I don't want this merged until we have agreed what the plan forward for #16261 is, as otherwise merging this cleanup will leave that PR in a limbo.

The standard solution for #16261 (use summary.ref) has already been repeatedly explained.

**Comment by ejgallego** [comment]
> The standard solution for #16261 (use summary.ref) has already been repeatedly explained.

Yup, and I think that's problematic, see my answer there, and a lot of questions I posed were not answered.

In particular, I don't think it is a good idea to introduce a public API for locating things based on the summary (which I deem wrong) just to remove it later on.

Again, what is the problem you have with discussing about this?

**Comment by gares** [comment]
There was enough discussion IMO for the removal of a hack, and the discussion goes nowhere since everybody is repeating the same arguments, which had already been discussed before.

We have 5 months or so to revert this if needed before it gets released, and as much time as we want to discuss your CEP. 

@coqbot merge now

**Comment by ejgallego** [comment]
Well this sets precedent for:
- removing discussion tags unilaterally
- ignoring  review feedback requesting changes 

Noted then.

**Review by ejgallego — COMMENTED** [review]
As noted in the discussion I am not ok to merge this without discussing the interaction with prior PRs. Moreover I don't share the protocol violation and lack of team spirit displayed here.

**Pull request merged**
